site/server/app/com/fortysevendeg/exercises/services/free/userprogressops.scala

```diff
@@ -28,137 +34,132 @@ final case class UpdateUserProgress(
     Free.inject[UserProgressOp, F](UpdateUserProgress(userProgress))
 
+  def getExerciseEvaluations(user: User, library: String, section: String): Free[F, List[UserProgress]] =
+    Free.inject[UserProgressOp, F](GetExerciseEvaluations(user, library, section))
+
+  def getSolvedExerciseCount(user: User, library: String, section: String): Free[F, Int] = for {
+    tried ← getExerciseEvaluations(user, library, section)
+  } yield tried.filter(_.succeeded).size
+
   def fetchMaybeUserProgress(user: Option[User]): Free[F, OverallUserProgress] = {
     user.fold(anonymousUserProgress)(fetchUserProgress)
   }
 
-  private[this] def anonymousUserProgress: Free[F, OverallUserProgress] =
-    EO.getLibraries.map({ libs ⇒
-      val libraries = libs.map(l ⇒ {
-        OverallUserProgressItem(
-          libraryName = l.name,
-          sections = 0,
-          totalSections = l.sections.size,
-          completed = false
-        )
-      })
-      OverallUserProgress(libraries = libraries)
+  private[this] def anonymousUserProgress: Free[F, OverallUserProgress] = for {
+    libraries ← EO.getLibraries
+    libs = libraries.map(l ⇒ {
+      OverallUserProgressItem(
+        libraryName = l.name,
+        completedSections = 0,
+        totalSections = l.sections.size
+      )
     })
+  } yield OverallUserProgress(libraries = libs)
+
+  def getCompletedSectionCount(user: User, library: Library): Free[F, Int] = {
+    for {
+      finishedSections ← Free.freeMonad[F].sequence(
+        library.sections.map(s ⇒ {
+          for {
+            solvedExercises ← getSolvedExerciseCount(user, library.name, s.name)
+          } yield solvedExercises == s.exercises.size
+        })
+      )
+    } yield finishedSections.filter(_ == true).size
+  }
 
   def fetchUserProgress(user: User): Free[F, OverallUserProgress] = {
-    import ConnectionIOOps._
     for {
-      lbs ← UPR.findByUserIdAggregated(user.id).liftF[F]
-      libNames = lbs.map(_._1)
-      completedSectionsByLibrary ← Free.freeMonad[F].sequence(
-        libNames map { name ⇒ UPR.completedSectionsByLibrary(user.id, name).liftF[F] }
-      ).map(counts ⇒ (libNames zip counts).toMap)
-      items = lbs map {
-        case (libraryName, _, succeeded) ⇒
-          EO
-            .getLibrary(libraryName)
-            .map(_ map (_.sections.size) getOrElse 0)
-            .map { totalSections ⇒
-              val completedSections = completedSectionsByLibrary.get(libraryName).getOrElse(0)
-              OverallUserProgressItem(
-                libraryName = libraryName,
-                sections = completedSections,
-                totalSections = totalSections,
-                completed = succeeded && completedSections == totalSections
-              )
-            }
-      }
-      list ← Free.freeMonad[F].sequence(items)
-    } yield OverallUserProgress(list)
+      allLibraries ← EO.getLibraries
+      libraryProgress ← Free.freeMonad[F].sequence(allLibraries.map(l ⇒ {
```

Comment:
Instead of explicitly using the `Free` monad instance here should we require this via implicit resolution in case there is a user provided instance for `Monad[Free[F, ?]]`?
This would be a good time to introduce Kind Projector too.